src/telnetio/_machine.py

```diff
@@ -26,245 +22,263 @@ class TelnetOption:
     reply_pending: bool = False
 
 
+class State(Enum):
+    #: Receiving normal data
+    data = auto()
+    #: Receiving data after CR received
+    data_cr = auto()
+    #: IAC
+    command = auto()
+    #: IAC with multiple (3) bytes
+    negotiation = auto()
+    #: Subnegotation option
+    subn_option = auto()
+    #: Subnegotation sub-opion
+    subn_suboption = auto()
+    #: Subnegotation data
+    subn_data = auto()
+    #: Subnegotation end
+    subn_end = auto()
+
+
+class ErrorKind(Enum):
+    SE_BUFFER_EMPTY = auto()
+    SE_BUFFER_NUL = auto()
+    SE_BUFFER_TOO_SHORT = auto()
+    SB_INVALID = auto()
+    OTHER = auto()
+
+
 class Event(ABC):
     @abstractmethod
     def as_bytes(self) -> bytes:  # pragma: nocover
         ...
 
 
+@dataclass(frozen=True)
+class Error(Event):
+    kind: ErrorKind
+    data: bytes = b""
+
+    def as_bytes(self) -> bytes:  # pragma: nocover
+        raise NotImplementedError()
+
+
+@dataclass(frozen=True)
+class Data(Event):
+    msg: bytes
+
+    def as_bytes(self) -> bytes:
+        return self.msg
+
+
 @dataclass(frozen=True)
 class Command(Event):
-    cmd: Opt
+    cmd: int
```

Comment:
Not something i expect you to address, but i do prefer this being typed to make it more clear that there are restrictions on the valid command values.

Reply:
Yeah, it was that way, but then you'd get a `ValueError` or equivalent if you received a weird byte on the wire, which is not great.  I think it's better if the logic handles the cmd value regardless.